File: src/contract_manager.py
"""
BSE Contract Manager - Daily Contract File Management
======================================================

Manages daily BSE contract files:
1. Fetches yesterday's contract file (API doesn't have today's file)
2. Deletes old files (older than 2 days)
3. Checks if file exists before downloading
4. Provides latest contract file to decoder

Example:
    Today: 27-11-2025
    Download: 26-11-2025 (yesterday)
    Keep: 26-11-2025, 25-11-2025 (2 days)
    Delete: 24-11-2025 and older
"""

import os
import logging
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional
from bse_contract_fetcher import BSEContractFetcher

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class BSEContractManager:
    """
    Manages daily BSE contract files with automatic cleanup.
    """
    
    def __init__(self, api_url: str, storage_dir: str = None):
        """
        Initialize contract manager.
        
        Args:
            api_url: BSE API endpoint
            storage_dir: Directory to store contract files (default: data/tokens/)
        """
        self.api_url = api_url
        self.fetcher = BSEContractFetcher(api_url)
        
        # Set storage directory
        if storage_dir is None:
            self.storage_dir = Path(__file__).parent.parent / 'data' / 'tokens'
        else:
            self.storage_dir = Path(storage_dir)
        
        # Create directory if not exists
        self.storage_dir.mkdir(parents=True, exist_ok=True)
        
        # Contracts dictionary - populated after loading
        self.contracts = {}
        
        logger.info(f"Contract Manager initialized")
        logger.info(f"Storage: {self.storage_dir}")
    
    def get_file_path(self, date_str: str) -> Path:
        """
        Get file path for a date.
        
        Args:
            date_str: Date in DD-MM-YYYY format
        
        Returns:
            Path to contract file
        """
        file_date = date_str.replace("-", "")
        filename = f"BSE_EQD_CONTRACT_{file_date}_fetched.csv"
        return self.storage_dir / filename
    
    def file_exists(self, date_str: str) -> bool:
        """
        Check if contract file exists for a date.
        
        Args:
            date_str: Date in DD-MM-YYYY format
        
        Returns:
            True if file exists
        """
        file_path = self.get_file_path(date_str)
        exists = file_path.exists()
        
        if exists:
            logger.info(f"✅ File exists: {file_path.name}")
        else:
            logger.info(f"❌ File missing: {file_path.name}")
        
        return exists
    
    def fetch_contract_file(self, date_str: str) -> bool:
        """
        Fetch contract file for a date.
        
        Args:
            date_str: Date in DD-MM-YYYY format
        
        Returns:
            True if successful
        """
        file_path = self.get_file_path(date_str)
        
        try:
            logger.info(f"📥 Fetching contract file for {date_str}...")
            
            # Fetch from API
            csv_content = self.fetcher.fetch_latest_csv(date_str)
            
            # Save to file
            self.fetcher.save_csv(csv_content, str(file_path))
            
            # Parse to verify
            self.fetcher.parse_csv_content(csv_content)
            
            logger.info(f"✅ Downloaded: {file_path.name}")
            logger.info(f"   Contracts: {self.fetcher.stats['total_contracts']}")
            
            return True
            
        except Exception as e:
            logger.error(f"❌ Failed to fetch {date_str}: {e}")
            return False
    
    def delete_old_files(self, keep_days: int = 2):
        """
        Delete contract files older than specified days.
        
        Args:
            keep_days: Number of recent days to keep (default: 2)
        """
        logger.info(f"🗑️  Cleaning up old files (keeping last {keep_days} days)...")
        
        # Calculate cutoff date
        cutoff_date = datetime.now() - timedelta(days=keep_days)
        
        # Find all contract files
        pattern = "BSE_EQD_CONTRACT_*_fetched.csv"
        files = list(self.storage_dir.glob(pattern))
        
        deleted_count = 0
        
        for file_path in files:
            try:
                # Extract date from filename
                # Format: BSE_EQD_CONTRACT_DDMMYYYY_fetched.csv
                filename = file_path.stem  # Remove .csv
                date_part = filename.split('_')[3]  # Get DDMMYYYY
                
                # Parse date
                file_date = datetime.strptime(date_part, '%d%m%Y')
                
                # Check if older than cutoff
                if file_date < cutoff_date:
                    logger.info(f"   Deleting: {file_path.name} ({file_date.strftime('%d-%m-%Y')})")
                    file_path.unlink()
                    deleted_count += 1
                else:
                    logger.info(f"   Keeping: {file_path.name} ({file_date.strftime('%d-%m-%Y')})")
                    
            except Exception as e:
                logger.warning(f"   Error processing {file_path.name}: {e}")
        
        if deleted_count > 0:
            logger.info(f"✅ Deleted {deleted_count} old file(s)")
        else:
            logger.info(f"✅ No old files to delete")
    
    def get_latest_file(self) -> Optional[Path]:
        """
        Get the latest available contract file.
        
        Returns:
            Path to latest file or None
        """
        # Find all contract files
        pattern = "BSE_EQD_CONTRACT_*_fetched.csv"
        files = list(self.storage_dir.glob(pattern))
        
        if not files:
            logger.warning("❌ No contract files found")
            return None
        
        # Sort by filename (DDMMYYYY is sortable)
        latest_file = sorted(files)[-1]
        
        logger.info(f"📂 Latest file: {latest_file.name}")
        return latest_file
    
    def update_contracts(self, date_str: str = None, keep_days: int = 2) -> bool:
        """
        Main update function - fetch contract file and cleanup.
        
        Process:
        1. Calculate target date (default: yesterday, API doesn't have today's file)
        2. Check if file exists
        3. Download if not exists
        4. Delete old files (older than keep_days)
        
        Args:
            date_str: Date in DD-MM-YYYY format (default: yesterday)
            keep_days: Number of days to keep files (default: 2)
        
        Returns:
            True if update successful
        """
        logger.info("=" * 80)
        logger.info("DAILY CONTRACT UPDATE")
        logger.info("=" * 80)
        
        # Calculate target date (yesterday if not specified)
        today = datetime.now()
        if date_str is None:
            yesterday = today - timedelta(days=1)
            target_date_str = yesterday.strftime("%d-%m-%Y")
        else:
            target_date_str = date_str
        
        logger.info(f"Today: {today.strftime('%d-%m-%Y')}")
        logger.info(f"Fetching: {target_date_str}")
        
        # Check if file exists
        if self.file_exists(target_date_str):
            logger.info("✅ File already exists - skipping download")
            success = True
        else:
            # Download contract file
            success = self.fetch_contract_file(target_date_str)
        
        if success:
            # Delete old files
            self.delete_old_files(keep_days=keep_days)
            
            logger.info("=" * 80)
            logger.info("✅ CONTRACT UPDATE COMPLETE")
            logger.info("=" * 80)
        else:
            logger.error("=" * 80)
            logger.error("❌ CONTRACT UPDATE FAILED")
            logger.error("=" * 80)
        
        return success
    
    def load_latest_contracts(self):
        """
        Load the latest contract file into memory.
        
        Returns:
            BSEContractFetcher with loaded contracts
        """
        latest_file = self.get_latest_file()
        
        if latest_file is None:
            raise FileNotFoundError("No contract files available")
        
        logger.info(f"📖 Loading contracts from: {latest_file.name}")
        
        # Read file
        with open(latest_file, 'r', encoding='utf-8') as f:
            csv_content = f.read()
        
        # Parse into fetcher
        self.fetcher.parse_csv_content(csv_content)
        
        # Populate contracts dict for token_mapper compatibility
        self.contracts = self.fetcher.contracts.copy()
        
        logger.info(f"✅ Loaded {self.fetcher.stats['total_contracts']} contracts")
        
        return self.fetcher


# Example usage
if __name__ == '__main__':
    print("\n" + "=" * 80)
    print("BSE CONTRACT MANAGER - DAILY UPDATE TEST")
    print("=" * 80)
    
    # Initialize manager
    api_url = "http://192.168.102.166:2060/v1/sftp-files"
    manager = BSEContractManager(api_url)
    
    # Run daily update
    success = manager.update_contracts()
    
    if success:
        print("\n" + "=" * 80)
        print("LOADING LATEST CONTRACTS")
        print("=" * 80)
        
        # Load latest contracts
        fetcher = manager.load_latest_contracts()
        
        # Test with sample tokens
        print("\n📋 TESTING TOKEN LOOKUP")
        print("-" * 80)
        
        test_tokens = [880622, 885255, 885890, 1117858]
        
        for token in test_tokens:
            contract = fetcher.get_contract(token)
            if contract:
                print(f"\n✅ Token {token}:")
                print(f"   {contract['symbol']:<10} {contract['full_name']}")
                if contract['expiry']:
                    print(f"   Expiry: {contract['expiry']:<15} Strike: ₹{contract['strike']:>10,.0f}")
                    print(f"   Type: {contract['option_type']:<4} Lot: {contract['lot_size']}")
            else:
                print(f"\n❌ Token {token}: NOT FOUND")
        
        print("\n" + "=" * 80)
        print("✅ CONTRACT MANAGER TEST COMPLETE")
        print("=" * 80)

File: bse-go/pkg/contract_manager/contract_manager.go
// Package contract_manager manages F&O Contract Master files.
// It auto-downloads from BSE API and cleans up old files.
package contract_manager

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"bse-go/pkg/contract_fetcher"
)

// ContractInfo holds F&O contract details
type ContractInfo struct {
	Token      string
	Symbol     string // TckrSymb
	Underlying string // underlying from symbol
	InstType   string // instrument type (IF, IO, SF, SO)
	Expiry     string // XpryDt
	ExpiryDisp string // formatted expiry for display
	Strike     string // StrkPric
	OptionType string // OptnTp (CE/PE)
	LotSize    string // MinLot
}

// Manager handles F&O contract file management
type Manager struct {
	fetcher   *contract_fetcher.Fetcher
	dataDir   string
	contracts map[string]ContractInfo // token → info
}

// NewManager creates a new contract manager
func NewManager(apiURL, dataDir string) *Manager {
	if dataDir == "" {
		dataDir = "data/tokens"
	}
	return &Manager{
		fetcher:   contract_fetcher.NewFetcher(apiURL),
		dataDir:   dataDir,
		contracts: make(map[string]ContractInfo),
	}
}

// UpdateContracts fetches latest contract file and cleans up old ones
func (m *Manager) UpdateContracts(date time.Time, keepDays int) error {
	// Build API path
	path, fileName := contract_fetcher.BuildFOPath(date)
	outputPath := filepath.Join(m.dataDir, strings.Replace(fileName, ".csv", "_fetched.csv", 1))

	// Check if file already exists
	if _, err := os.Stat(outputPath); err == nil {
		log.Printf("✅ Contract file already exists: %s", outputPath)
	} else {
		// Download new file
		log.Printf("📥 Downloading F&O contract file: %s/%s", path, fileName)
		if err := m.fetcher.FetchAndSave(path, fileName, outputPath); err != nil {
			return fmt.Errorf("failed to download contract file: %w", err)
		}
		log.Printf("✅ Downloaded: %s", outputPath)
	}

	// Cleanup old files
	m.deleteOldFiles(keepDays)

	// Load contracts into memory
	return m.loadContracts(outputPath)
}

// LoadFromCache loads contracts from existing CSV files without API fetch
func (m *Manager) LoadFromCache() error {
	// Try multiple patterns to find contract files
	patterns := []string{
		filepath.Join(m.dataDir, "BSE_EQD_CONTRACT_*_fetched.csv"),
		filepath.Join(m.dataDir, "BSE_EQD_CONTRACT_*.csv"),
	}

	var allMatches []string
	for _, pattern := range patterns {
		matches, err := filepath.Glob(pattern)
		if err != nil {
			continue
		}
		allMatches = append(allMatches, matches...)
	}

	if len(allMatches) == 0 {
		return fmt.Errorf("no contract files found in %s", m.dataDir)
	}

	// Sort to get latest (by filename date)
	sort.Strings(allMatches)
	latestFile := allMatches[len(allMatches)-1]

	log.Printf("📂 Loading contracts from: %s", latestFile)
	return m.loadContracts(latestFile)
}

// loadContracts parses CSV and loads into memory
func (m *Manager) loadContracts(csvPath string) error {
	file, err := os.Open(csvPath)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return fmt.Errorf("failed to read CSV: %w", err)
	}

	if len(records) < 2 {
		return fmt.Errorf("CSV file is empty or has no data rows")
	}

	// Build column index map from header
	header := records[0]
	colIndex := make(map[string]int)
	for i, col := range header {
		colIndex[strings.TrimSpace(col)] = i
	}

	// Required columns for F&O
	// FinInstrmId: token
	// TckrSymb: symbol (e.g., SENSEX25JULFUT, SENSEX25JUL85000CE)
	// XpryDt: expiry date
	// StrkPric: strike price
	// OptnTp: option type (CE/PE/XX)
	// MinLot: lot size

	tokenIdx, ok1 := colIndex["FinInstrmId"]
	symbolIdx, ok2 := colIndex["TckrSymb"]
	expiryIdx, ok3 := colIndex["XpryDt"]
	strikeIdx, ok4 := colIndex["StrkPric"]
	optionIdx, ok5 := colIndex["OptnTp"]
	lotIdx, ok6 := colIndex["MinLot"]

	if !ok1 || !ok2 {
		return fmt.Errorf("required columns (FinInstrmId, TckrSymb) not found in CSV")
	}

	// Clear existing contracts
	m.contracts = make(map[string]ContractInfo)

	// Parse data rows
	for i, row := range records[1:] {
		if len(row) <= tokenIdx || len(row) <= symbolIdx {
			continue
		}

		token := strings.TrimSpace(row[tokenIdx])
		symbol := strings.TrimSpace(row[symbolIdx])

		if token == "" || symbol == "" {
			continue
		}

		info := ContractInfo{
			Token:  token,
			Symbol: symbol,
		}

		// Extract optional fields
		if ok3 && len(row) > expiryIdx {
			info.Expiry = strings.TrimSpace(row[expiryIdx])
			info.ExpiryDisp = formatExpiry(info.Expiry)
		}
		if ok4 && len(row) > strikeIdx {
			info.Strike = strings.TrimSpace(row[strikeIdx])
		}
		if ok5 && len(row) > optionIdx {
			info.OptionType = strings.TrimSpace(row[optionIdx])
		}
		if ok6 && len(row) > lotIdx {
			info.LotSize = strings.TrimSpace(row[lotIdx])
		}

		// Parse underlying and instrument type from symbol
		info.Underlying, info.InstType = parseSymbol(symbol)

		m.contracts[token] = info

		// Log first few for verification
		if i < 3 {
			log.Printf("   [%d] Token %s → %s (%s %s %s)",
				i+1, token, info.Underlying, info.OptionType, info.Strike, info.ExpiryDisp)
		}
	}

	log.Printf("✅ Loaded %d F&O contracts", len(m.contracts))
	return nil
}

// deleteOldFiles removes contract files older than keepDays
func (m *Manager) deleteOldFiles(keepDays int) {
	pattern := filepath.Join(m.dataDir, "BSE_EQD_CONTRACT_*_fetched.csv")
	matches, err := filepath.Glob(pattern)
	if err != nil {
		log.Printf("⚠️ Failed to glob for cleanup: %v", err)
		return
	}

	if len(matches) <= keepDays {
		return // Nothing to delete
	}

	// Sort by filename (contains date)
	sort.Strings(matches)

	// Delete oldest files, keep last 'keepDays' files
	toDelete := matches[:len(matches)-keepDays]
	for _, f := range toDelete {
		if err := os.Remove(f); err != nil {
			log.Printf("⚠️ Failed to delete %s: %v", f, err)
		} else {
			log.Printf("🗑️ Deleted old file: %s", filepath.Base(f))
		}
	}
}

// GetContract returns contract info for a token
func (m *Manager) GetContract(token string) (ContractInfo, bool) {
	info, ok := m.contracts[token]
	return info, ok
}

// GetContracts returns all contracts
func (m *Manager) GetContracts() map[string]ContractInfo {
	return m.contracts
}

// Count returns number of loaded contracts
func (m *Manager) Count() int {
	return len(m.contracts)
}

// parseSymbol extracts underlying and instrument type from F&O symbol
// Examples:
//
//	SENSEX25JULFUT → SENSEX, FUT
//	SENSEX25JUL85000CE → SENSEX, CE
//	BANKEX25NOV57000PE → BANKEX, PE
func parseSymbol(symbol string) (underlying, instType string) {
	// Common underlyings in BSE F&O
	underlyings := []string{"SENSEX", "BANKEX", "SENSEX50"}

	for _, u := range underlyings {
		if strings.HasPrefix(symbol, u) {
			// Check for FUT
			if strings.Contains(symbol, "FUT") {
				return u, "FUT"
			}
			// Check for CE/PE at end
			if strings.HasSuffix(symbol, "CE") {
				return u, "CE"
			}
			if strings.HasSuffix(symbol, "PE") {
				return u, "PE"
			}
			return u, "UNKNOWN"
		}
	}

	// Try regex for other symbols
	// Pattern: (SYMBOL)(YYMMMM)(STRIKE)?(CE|PE|FUT)
	re := regexp.MustCompile(`^([A-Z]+)(\d{2}[A-Z]{3})`)
	matches := re.FindStringSubmatch(symbol)
	if len(matches) >= 2 {
		underlying = matches[1]
		if strings.Contains(symbol, "FUT") {
			return underlying, "FUT"
		}
		if strings.HasSuffix(symbol, "CE") {
			return underlying, "CE"
		}
		if strings.HasSuffix(symbol, "PE") {
			return underlying, "PE"
		}
	}

	return symbol, ""
}

// formatExpiry converts expiry date to display format
// Input: 2025-09-18 or 18-Sep-2025
// Output: 18-Sep-2025
func formatExpiry(expiry string) string {
	// Try different date formats
	formats := []string{
		"2006-01-02",
		"02-01-2006",
		"02-Jan-2006",
	}

	for _, fmt := range formats {
		if t, err := time.Parse(fmt, expiry); err == nil {
			return t.Format("02-Jan-2006")
		}
	}

	return expiry // Return as-is if can't parse
}

// BuildSymbolDisplay creates a display-friendly symbol
func (m *Manager) BuildSymbolDisplay(token string) string {
	info, ok := m.contracts[token]
	if !ok {
		return fmt.Sprintf("TOKEN_%s", token)
	}

	// Futures: SENSEX_FUT_18-Sep-2025
	if info.InstType == "FUT" || strings.Contains(strings.ToUpper(info.OptionType), "XX") {
		return fmt.Sprintf("%s_FUT_%s", info.Underlying, info.ExpiryDisp)
	}

	// Options: SENSEX_CE_85000_18-Sep-2025
	if info.OptionType != "" && info.Strike != "" {
		// Clean up strike (remove decimals if .00)
		strike := info.Strike
		if f, err := strconv.ParseFloat(strike, 64); err == nil {
			if f == float64(int(f)) {
				strike = fmt.Sprintf("%.0f", f)
			}
		}
		return fmt.Sprintf("%s_%s_%s_%s", info.Underlying, info.OptionType, strike, info.ExpiryDisp)
	}

	return info.Symbol
}
